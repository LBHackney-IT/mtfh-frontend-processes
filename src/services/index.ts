export { default as locale } from "./locale";
export { default as menu } from "./menu";
export * as processes from "./processes";
export * from "./config";
