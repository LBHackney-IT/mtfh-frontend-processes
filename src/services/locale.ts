import reactHtmlParser from "react-html-parser";

import { AssetAddress } from "@mtfh/common/lib/api/asset/v1";

const locale = {
  backButton: "Back",
  bookAppointment: "Book Appointment",
  next: "Next",
  cancel: "Cancel",
  cancelProcess: "Cancel process",
  change: "Change",
  views: {
    soleToJoint: {
      title: "Sole tenant requests a joint tenure",
      steps: {
        selectTenant: "Select person",
        checkEligibility: "Eligibility Checks",
        supportingDocuments: "Supporting documents",
        finish: "Finish",
        breachOfTenancy: "Breach of tenure check",
        requestDocuments: "Request Documents",
        reviewDocuments: "Review Documents",
        submitCase: "Submit case",
      },
      actions: {
        reassignCase: "Reassign case",
        cancelProcess: "Cancel process",
        caseActivityHistory: "Case activity history",
      },
    },
    selectTenants: {
      selectTenantHint:
        "This person will be asked for proof of relationship e.g. marriage or civil partnership certificate.",
      selectTenantLabel: "Who do you want to add as a joint tenant?",
      addToTenureText: "If the person you want to add is not listed you must first",
      addToTenureLink: "add them to the tenure",
      noHouseholdMembersOver18:
        "There are no household members over the age of 18 associated to this.",
    },
    processesMenu: {
      selectProcessDescription: () =>
        reactHtmlParser(
          "<p>Start a process by selecting a process category below. <br /> Where a process is not yet supported in <strong>Manage My Home</strong> these links will open the appropriate interim Google form.</p>",
        ),
    },
    checkEligibility: {
      autoCheckIntro:
        "Manage My Home has used tenure, person and property information to automatically check eligibility for this application.",
      passedChecks: "Passed automatic eligibilty checks",
      failedChecks: "Not eligible for a sole to joint tenure",
    },
    breachOfTenancy: {
      autoCheckIntro:
        "Manage My Home has used tenure, person and property information to automatically check eligibility for this application.",
      passedChecks: "Eligibility checks passed",
      failedChecks: "Not eligible for a sole to joint tenure",
    },
    reviewDocuments: {
      passedChecks: "Eligibility checks passed",
      documentsRequested: "Documents requested",
      checkSupportingDocumentsAppointment:
        "I need to make an appointment to check supporting documents",
      appointmentScheduled: "Office appointment scheduled",
      seenPhotographicId:
        "I confirm that an outcome letter has been sent to the resident",
      seenSecondId:
        "I confirm I have seen a second form of ID (does not have to be photographic)",
      isNotInImmigrationControl:
        "I confirm that the prospective tenant is not subject to immigration control under the Asylum And Immigration Act 1996",
      seenProofOfRelationship:
        "I confirm I have seen proof of relationship to the existing tenant",
      incomingTenantLivingInProperty:
        "I confirm I have seen 3 separate documents proving the proposed tenant has been living at the property for a minimum of 12 months",
    },
  },
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    unableToFindProcess: "Invalid Process",
    unableToFindProcessDescription: "We have been unable to find a process of this type.",
    unableToFindState: "Invalid State",
    unableToFindStateDescription: "We have been unable to process a state of this type.",
  },
  title: "Processes",
  loadingText: "Submitting...",
  components: {
    startProcess: {
      buttonLabel: "Start process",
      thirdPartyHeading: "Sharing information with third parties",
    },
    entitySummary: {
      tenurePaymentRef: "Tenure payment ref",
      address: (assetAddress: AssetAddress): string => {
        const {
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,
          addressLine4,
          postCode,
        } = assetAddress;

        return [
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,

          addressLine4,
          postCode,
        ]
          .filter((addressLine) => !!addressLine)
          .join(" ");
      },
    },
  },
};

export default locale;
