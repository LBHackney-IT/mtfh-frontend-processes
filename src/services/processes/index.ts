export * from "./sole-to-joint";
export * from "./change-of-name";
