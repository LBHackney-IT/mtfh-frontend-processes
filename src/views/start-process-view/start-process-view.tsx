import { useState } from "react";
import { Link as RouterLink, useParams } from "react-router-dom";

import { EntitySummary, StartProcess } from "../../components";
import { locale, processes } from "../../services";
import { IProcess, TargetType } from "../../types";

import { Asset } from "@mtfh/common/lib/api/asset/v1/types";
import { Patch } from "@mtfh/common/lib/api/patch/v1/types";
import { RelatedEntity } from "@mtfh/common/lib/api/process/v1";
import { ErrorSummary, Layout, Link } from "@mtfh/common/lib/components";

interface ParamProps {
  targetId: string;
  targetType: TargetType;
  processName: string;
}

export const StartProcessView = () => {
  const { targetId, targetType, processName } = useParams<ParamProps>();
  const [relatedEntities, setRelatedEntities] = useState<RelatedEntity[]>([]);
  const [asset, setAsset] = useState<Asset>();
  const [patch, setPatch] = useState<Patch>();

  const SideBar = () => {
    return null;
  };

  const process = Object.values(processes).find(
    (process) => process.processName === processName,
  );

  if (!process) {
    return (
      <ErrorSummary
        id="start-process-view"
        title={locale.errors.unableToFindProcess}
        description={locale.errors.unableToFindProcessDescription}
      />
    );
  }

  const { startProcess, title }: IProcess = process;

  const backLink = `/processes/${targetType}/${targetId}`;

  return (
    <Layout
      data-testid={`${processName}-start`}
      sidePosition="right"
      backLink={
        <Link as={RouterLink} to={backLink} variant="back-link">
          {locale.backButton}
        </Link>
      }
      top={<h1 className="lbh-heading-h1">{title}</h1>}
      side={<SideBar />}
    >
      <>
        <EntitySummary
          type={targetType}
          id={targetId}
          setRelatedEntities={setRelatedEntities}
          setAsset={setAsset}
          setPatch={setPatch}
        />
        {asset && relatedEntities && patch && (
          <StartProcess
            targetId={targetId}
            processName={processName}
            process={startProcess}
            backLink={backLink}
            targetType={targetType}
            relatedEntities={relatedEntities}
            patches={patch}
            asset={asset}
          />
        )}
      </>
    </Layout>
  );
};
