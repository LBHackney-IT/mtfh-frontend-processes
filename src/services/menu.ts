import * as processes from "./processes";

import { Asset } from "@mtfh/common/lib/api/asset/v1/types";
import { Tenure } from "@mtfh/common/lib/api/tenure/v1/types";

interface MenuProps {
  label: string;
  link: string;
  processes?: { label: string; link: string; targetType: string }[];
  getPplQuery?: (context: any) => string;
}

const getProcessLink = (processName: string) => {
  const process = processes[processName];
  return {
    targetType: process.targetType,
    label: process.title,
    link: `/processes/${process.processName}/start`,
  };
};

const getPplQueryForRequestALegalReferral = ({ tenure, asset }): string => {
  if (!tenure || !asset) {
    return "";
  }
  const {
    householdMembers,
    tenuredAsset: { propertyReference },
  } = tenure as Tenure;
  const {
    assetAddress: { addressLine1, postCode },
  } = asset as Asset;
  const tenantNames = householdMembers
    .filter((member) => member.isResponsible)
    .map((member) => member.fullName)
    .join(", ");
  const address = addressLine1;
  const query = `usp=pp_url&entry.1699199605=${tenantNames}&entry.2086190845=${address}&entry.1956036523=${postCode}&entry.1140527432=${propertyReference}`;
  const encodedQuery = encodeURI(query);
  return encodedQuery;
};

const getPplQueryTenancyChange = ({ tenure }): string => {
  if (!tenure) {
    return "";
  }
  const {
    householdMembers,
    paymentReference,
    tenuredAsset: { fullAddress, propertyReference },
  } = tenure as Tenure;

  const tenant = householdMembers.find((member) => member.isResponsible);
  const query = `usp=pp_url&entry.1942863541=${tenant?.fullName}&entry.1090481808=${fullAddress}&entry.321634337=${propertyReference}&entry.582205322=${paymentReference}`;
  const encodedQuery = encodeURI(query);
  return encodedQuery;
};

const menu: MenuProps[] = [
  {
    label: "Sole to Joint",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSdgJ9DSgGI0Aj7GO1bzLbbrArPabjS8DQwmvwb9ltB-qYYESA/viewform",
    processes: [
      getProcessLink(processes.soletojoint.processName),
      getProcessLink(processes.changeOfName.processName),
    ],
  },
  {
    label: "Other tenancy changes",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSdgJ9DSgGI0Aj7GO1bzLbbrArPabjS8DQwmvwb9ltB-qYYESA/viewform",
    getPplQuery: getPplQueryTenancyChange,
  },
  {
    label: "Record a tenancy / audit homecheck",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfyXscVvgCBhbBBe7ADpV6cUqJoKrrTbU0YYRB_uXS2J4AHMQ/viewform",
  },
  {
    label: "Find legal referral response",
    link: "https://docs.google.com/spreadsheets/d/1fd6FwBXvBo2DmX8qZaXUr7hX9sE6Zr7HZsPlA_OO8Hs/edit#gid=1251146140",
  },
  {
    label: "Request a legal referral",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScpHsoJQgq112HIMbW06YAFpggijwQuODJM483HGbX_PMfCMw/viewform",
    getPplQuery: getPplQueryForRequestALegalReferral,
  },
  {
    label: "Record notice served",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfRSMia4FeT6WPjrmRlsLpOXyJ4tFs3Sv_O0toxUbv_FmIjKw/viewform",
  },
  {
    label: "Rehousing",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScPXSs4-xxTH2N5dUwtfmEl5mpu7kxZ6mHe86Qy_rlDNu-GbA/viewform",
  },
  {
    label: "Person",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScxrWFLF7M7A3B1yzY-LZFXkV0P0qv6igcQ4wEgWHQk8DaUCw/viewform",
  },
  {
    label: "Report Japanese knotweed",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSceU2H-VWi2ewaEiOAkOWULVFU2DoPSen8SRPyJrmQtRpEhfQ/viewform?gxids=7628",
  },
  {
    label: "Report overgrown tree",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfLmlLraDyOfqYhdjkOfkNWWlHUUep05G5n0GN-ewizN0Fbuw/viewform?gxids=7628",
  },
  {
    label: "Report overgrown garden",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSc3mOUUEQhdRiDsS0k9XDWlfGJG0CEFAC2rnC1sYmM-idt-Nw/viewform?gxids=7628",
  },
];

export default menu;
