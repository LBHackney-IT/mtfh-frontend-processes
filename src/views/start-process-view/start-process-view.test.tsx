import { render } from "@hackney/mtfh-test-utils";
import { screen, waitFor } from "@testing-library/react";

import { locale } from "../../services";
import { StartProcessView } from "./start-process-view";

test("it renders soletojoint start process view", async () => {
  const { container } = render(<StartProcessView />, {
    url: "/processes/soletojoint/start/tenure/8f6ff46b-6a92-0311-1ef6-61508321e65f",
    path: "/processes/:processName/start/:targetType/:targetId",
  });

  await expect(
    screen.findByText(locale.components.entitySummary.tenurePaymentRef, { exact: false }),
  ).resolves.toBeInTheDocument();
  await expect(
    screen.findByText(locale.components.startProcess.buttonLabel),
  ).resolves.toBeInTheDocument();
  await waitFor(() => {
    expect(container).toMatchSnapshot();
  });
});

test("it renders an error if an invalid process name is passed", async () => {
  render(<StartProcessView />, {
    url: "/processes/invalid-process-name/start/tenure/8f6ff46b-6a92-0311-1ef6-61508321e65f",
    path: "/processes/:processName/start/:targetType/:targetId",
  });

  await expect(
    screen.findByText(locale.errors.unableToFindProcess),
  ).resolves.toBeInTheDocument();
  await expect(
    screen.findByText(locale.errors.unableToFindProcessDescription),
  ).resolves.toBeInTheDocument();
});
