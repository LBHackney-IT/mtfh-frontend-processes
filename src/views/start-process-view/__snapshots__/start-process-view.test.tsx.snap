// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders soletojoint start process view 1`] = `
<div>
  <div
    class="mtfh-layout mtfh-layout--right"
    data-testid="soletojoint-start"
  >
    <a
      class="govuk-back-link lbh-back-link lbh-link--no-visited-state"
      href="/processes/tenure/8f6ff46b-6a92-0311-1ef6-61508321e65f"
    >
      Back
    </a>
    <div
      id="content"
    />
    <h1
      class="lbh-heading-h1"
    >
      Sole tenant requests a joint tenure
    </h1>
    <div
      class="mtfh-layout__container"
    >
      <div
        class="mtfh-layout__aside"
      />
      <div
        class="mtfh-layout__main"
      >
        <h2
          class="lbh-heading-h2 entity-summary__tenure-heading"
        >
          <div>
            Tenure payment ref
             
            mt0xzvjp6g
          </div>
          <div>
            72251 Champlin Trafficway, 99647
          </div>
          <div>
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state"
              href="/person/a302fe28-e5c3-4c84-a3e6-cd575e65093f"
            >
              Global Web Consultant Alverta Denesik
            </a>
          </div>
        </h2>
        <form
          action="#"
        >
          <h3>
            Sharing information with third parties
          </h3>
          <p
            class="lbh-body-m"
          >
            By starting this process the tenant and proposed tenant must agree that information relating to this application will be shared with third parties, for instance to perform a credit check.
          </p>
          <div
            class="govuk-checkboxes__item"
          >
            <input
              class="govuk-checkboxes__input"
              id="condition"
              name="condition"
              type="checkbox"
              value="false"
            />
            <label
              class="govuk-label govuk-checkboxes__label"
              for="condition"
            >
              I have explained to the tenant and proposed tenant that their information will be shared with third parties.
            </label>
          </div>
          <h3>
            Sole to joint tenure application risks
          </h3>
          <div>
            <p
              class="lbh-body-m"
            >
              A joint tenancy is when two or more adults (aged 18 or over) are named on the tenure agreement. Joint tenants have equal rights and responsibilities under the tenure agreement for the duration of the tenancy.
            </p>
            <p
              class="lbh-body-m"
            >
              Joint tenants are both responsible, together and individually, for keeping to these tenure conditions and payments to the London borough of Hackney. Applying for a joint tenancy is at the discretion of the Housing team.
            </p>
            <h4
              class="lbh-heading-h4"
            >
              Things to consider if joint tenants separate
            </h4>
            <p
              class="lbh-body-m"
            >
              If a joint tenants relationship ends, discussions should be made about what will happen to the tenure.
            </p>
            <p
              class="lbh-body-m"
            >
              The tenant and proposed tenant need to understand:
            </p>
            <ul
              class="lbh-list lbh-list--bullet"
            >
              <li>
                <p
                  class="lbh-body-m"
                >
                  When can they end a joint tenure
                </p>
              </li>
              <li>
                <p
                  class="lbh-body-m"
                >
                  What will happen if they end a joint tenure but one of the tenants remains at the property
                </p>
              </li>
            </ul>
            <p
              class="lbh-body-m"
            >
              This information is available on the
               
              <a
                class="govuk-link lbh-link"
                href="https://hackney.gov.uk/your-tenancy-agreement"
                rel="noopener noreferrer"
                target="_blank"
              >
                terms and conditions of the tenancy
              </a>
            </p>
          </div>
          <div
            class="start-process__actions"
          >
            <button
              class="govuk-button lbh-button lbh-button--chevron start-process__start-button"
              disabled=""
              type="submit"
            >
              Start process
              <svg
                fill="none"
                height="22"
                viewBox="0 0 20 22"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M0 0H8.54573L20 10H11.4543L0 0Z"
                  fill="white"
                  fill-rule="evenodd"
                />
                <path
                  clip-rule="evenodd"
                  d="M0 22L8.54573 22L20 10H11.4543L0 22Z"
                  fill="#96CCAE"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state start-process__cancel-link"
              href="/processes/tenure/8f6ff46b-6a92-0311-1ef6-61508321e65f"
            >
              Cancel
            </a>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
